bin/annotate_test_failures

```diff
@@ -14,31 +18,102 @@ unless File.exist?(junit_path)
+    "#{@classname}.#{@name}"
+  end
+
+  def ultimately_succeeds?(parent_node:)
+    all_same_test_nodes = parent_node.get_elements("testcase[@classname='#{@classname}' and @name='#{@name}']")
+    !all_same_test_nodes.last.elements['failure'] # If last node found for that test doesn't have a <failure> child, then test ultimately succeeded.
+  end
+
+  def ==(other)
+    @classname == other.classname && @name == other.name && @message == other.message && @details == other.details
+  end
+end
+
+# Add a test_failure into a given array, or increment its count if it already exists
+#
+def record(test_failure, into:)
+  existing = into.find { |f| f == test_failure }
+  if existing.nil?
+    into.append(test_failure)
+  else
+    existing.count += 1
+  end
+end
+
+# Given a list of failures for a given step and state, update the corresponding annotation
+#
+def update_annotation(title, list, style, state)
+  annotation_context = "#{title} (#{state})"
+  if list.empty?
+    puts "No test #{state}. Removing any previous `#{annotation_context}` Buildkite annotation if any.\n\n"
+    system('buildkite-agent', 'annotation', 'remove', '--context', annotation_context)
+  else
+    tests_count = list.map(&:key).uniq.count # Count the number of tests that failed, even if some tests might have multiple assertion failures
+    puts "#{tests_count} test(s) #{state} (#{list.count} distinct assertion failures in total). Reporting them as a `#{annotation_context}` Buildkite #{style} annotation.\n\n"
+    markdown = "\#\#\#\# #{title}: #{tests_count} tests #{state} (#{list.count} distinct assertion failure(s) in total)\n\n" + list.map(&:to_s).join("\n")
+    puts markdown
+    system('buildkite-agent', 'annotate', markdown, '--style', style, '--context', annotation_context)
+  end
+end
+
+###################
+# Main
+###################
+
 file = File.open(junit_path)
 xmldoc = REXML::Document.new(file)
 
 failures = []
-REXML::XPath.each(xmldoc, '//testcase[failure]') do |testcase|
-  failure = testcase.elements['failure']
-  failures.append <<~ENTRY
-  <details><summary><tt>#{testcase['name']}</tt> in <tt>#{testcase['classname']}</tt></summary>
-  
-  #{failure['message']}
-
-  ```
-  #{failure.text}
-  ```
-  </details>
-  ENTRY
-end
+flakies = []
+REXML::XPath.each(xmldoc, '//testcase[failure]') do |node|
+  test_failure = TestFailure.new(node)
 
-if failures.empty?
-  puts "No test failure found. Removing any previous `#{title}` Buildkite annotation if any."
-  system('buildkite-agent', 'annotation', 'remove', '--context', title)
-  exit 0 # We don't want to fail if the `annotation remove` command failed because there was no previous annotation
-else
-  puts "Found #{failures.count} test failure(s). Reporting them as a `#{title}` Buildkite annotation."
-  markdown = "\#\#\#\# #{title}: #{failures.length} failure(s)\n\n" + failures.join("\n")
-  system('buildkite-agent', 'annotate', markdown, '--style', 'error', '--context', title)
-  exit $?.exitstatus
```

Comment:
I wondered a bit about the explicit `exit` disappearing from the code, but I guess it no longer makes sense to bubble up the exit code of the `buildkite-agent annotate` command when we may call it more than once.

The important thing is that the test task reports the correct exit code, but that's already handled [by the way we capture the tests exit code before calling the annotation step](https://github.com/wordpress-mobile/WordPress-iOS/blob/1be5d8cf6be34a2245400a2164375d123fb161b6/.buildkite/commands/run-unit-tests.sh#L15-L38).

Reply:
Exactly my reasoning for why I removed it 🙂👍